pyo3-macros-backend/src/pyclass.rs

```diff
@@ -439,15 +447,39 @@ fn impl_class(
             fn for_each_method_def(visitor: &mut dyn FnMut(&pyo3::class::PyMethodDefType)) {
                 use pyo3::class::impl_::*;
                 let collector = PyClassImplCollector::<Self>::new();
-                #iter_py_methods
-                    .chain(collector.py_class_descriptors())
-                    .chain(collector.object_protocol_methods())
-                    .chain(collector.async_protocol_methods())
-                    .chain(collector.context_protocol_methods())
-                    .chain(collector.descr_protocol_methods())
-                    .chain(collector.mapping_protocol_methods())
-                    .chain(collector.number_protocol_methods())
-                    .for_each(visitor)
+                #for_each_py_method;
+
+                for method_def in collector.py_class_descriptors() {
+                    visitor(method_def);
+                }
+
+                for method_def in collector.object_protocol_methods() {
+                    visitor(method_def);
+                }
+
+                for method_def in collector.async_protocol_methods() {
+                    visitor(method_def);
+                }
+
+                for method_def in collector.context_protocol_methods() {
+                    visitor(method_def);
+                }
+
+                for method_def in collector.descr_protocol_methods() {
+                    visitor(method_def);
+                }
+
+                for method_def in collector.mapping_protocol_methods() {
+                    visitor(method_def);
+                }
+
+                for method_def in collector.number_protocol_methods() {
+                    visitor(method_def);
+                }
+
+                for method_def in collector.py_class_descriptors() {
+                    visitor(method_def);
+                }
```

Comment:
`collector.py_class_descriptors()` is visited twice here

Reply:
Nice catch!

I was wondering if we should build an iterator to avoid so much repetition of the boilerplate, to make errors like this easier to spot. Something like the below:

```rust

const IMPLS: &[syn::Ident] = &[syn::Ident::new("descr_protocol_methods", Span::call_site()), ...];

let impls = IMPLS.iter().map(|Ident| quote!(for method_def in collector.#ident() { ... }));

// use #impls in main output
```

Sorry for the abbreviated code, typing on my phone. Hopefully the gist comes through.

Reply:
@mejrs thanks for spotting, this is fixed

Reply:
I also have a patch where the visitor accept the slices directly instead of individual items, so we have:
```rust
visitor(collector.py_class_descriptors());
visitor(collector.object_protocol_methods());
visitor(collector.async_protocol_methods());
visitor(collector.context_protocol_methods());
visitor(collector.descr_protocol_methods());
visitor(collector.mapping_protocol_methods());
visitor(collector.number_protocol_methods());
```

This is closer to the previous implementation and IMO clearer. It's also smaller still in terms of llvm-lines, for a total of 45552 on the word-count example.